### ISSUE TITLE
Backup the history and starred commands

### ISSUE THREAD

**Issue by RicardoOrtizG** [OpenRefine/OpenRefine#3512]
After any migration of version the history and starred commands get deleted, if come back to the older version it still deleted, so get lost apparently. 

### Proposed solution
Documentation to save the history and commands saved in OpenRefine or a way to import automatically the command history when migrating from a newer version.

**Comment by wetneb** [comment]
Which versions did you migrate from and to?

**Comment by RicardoOrtizG** [comment]
It happened always, from 2.8 to 3.0 and from 3.0 to 3.4. I hope this helps

**Comment by thadguidry** [comment]
Cannot reproduce with newer versions of OpenRefine.
I just did a small test with no previous workspace and created 2 small projects to test what things look like when going from 3.3 to 3.5.2 and I had no issue seeing the starred or history in the Expression dialog.   My `workspace.json` and `workspace.old` indeed still had the scripting.expressions and scripting.starred-expressions :

```json
{
    "projectIDs": [
        2134256808527,
        2630730136395
    ],
    "preferences": {
        "entries": {
            "reconciliation.standardServices": [
                {
                    "name": "Wikidata (old) (en)",
                    "view": {
                        "url": "https://www.wikidata.org/wiki/{{id}}"
                    },
                    "identifierSpace": "http://www.wikidata.org/entity/",
                    "schemaSpace": "http://www.wikidata.org/prop/direct/",
                    "suggest": {
                        "type": {
                            "service_url": "https://openrefine-wikidata.toolforge.org",
                            "service_path": "/en/suggest/type",
                            "flyout_service_path": "/en/flyout/type?id=${id}"
                        },
                        "property": {
                            "service_url": "https://openrefine-wikidata.toolforge.org",
                            "service_path": "/en/suggest/property",
                            "flyout_service_path": "/en/flyout/property?id=${id}"
                        },
                        "entity": {
                            "service_url": "https://openrefine-wikidata.toolforge.org",
                            "service_path": "/en/suggest/entity",
                            "flyout_service_path": "/en/flyout/entity?id=${id}"
                        }
                    },
                    "preview": {
                        "url": "https://openrefine-wikidata.toolforge.org/en/preview?id={{id}}",
                        "width": 400,
                        "height": 100
                    },
                    "defaultTypes": [
                        {
                            "id": "Q35120",
                            "name": "entity"
                        }
                    ],
                    "extend": {
                        "propose_properties": {
                            "service_url": "https://openrefine-wikidata.toolforge.org",
                            "service_path": "/en/propose_properties"
                        },
                        "property_settings": [
                            {
                                "name": "limit",
                                "label": "Limit",
                                "help_text": "Maximum number of values to return per row (0 for no limit)",
                                "type": "number",
                                "default": 0
                            },
                            {
                                "name": "rank",
                                "label": "Ranks",
                                "help_text": "Filter statements by rank",
                                "type": "select",
                                "default": "best",
                                "choices": [
                                    {
                                        "value": "any",
                                        "name": "Any rank"
                                    },
                                    {
                                        "value": "best",
                                        "name": "Only the best rank"
                                    },
                                    {
                                        "value": "no_deprecated",
                                        "name": "Preferred and normal ranks"
                                    }
                                ]
                            },
                            {
                                "name": "references",
                                "label": "References",
                                "help_text": "Filter statements by their references",
                                "type": "select",
                                "default": "any",
                                "choices": [
                                    {
                                        "value": "any",
                                        "name": "Any statement"
                                    },
                                    {
                                        "value": "referenced",
                                        "name": "At least one reference"
                                    },
                                    {
                                        "value": "no_wiki",
                                        "name": "At least one non-wiki reference"
                                    }
                                ]
                            },
                            {
                                "name": "count",
                                "label": "Return counts instead of values",
                                "help_text": "The number of values will be returned.",
                                "type": "checkbox",
                                "default": false
                            }
                        ]
                    },
                    "url": "https://tools.wmflabs.org/openrefine-wikidata/en/api",
                    "ui": {
                        "handler": "ReconStandardServicePanel",
                        "access": "json"
                    }
                }
            ],
            "scripting.starred-expressions": {
                "top": 2147483647,
                "list": [
                    "grel:value + \"new\"",
                    "grel:value + \"test\""
                ],
                "class": "com.google.refine.preference.TopList"
            },
            "scripting.expressions": {
                "top": 100,
                "list": [
                    "grel:value",
                    "grel:value + \"new\"",
                    "grel:value + \"test\""
                ],
                "class": "com.google.refine.preference.TopList"
            }
        }
    }
}
```

**Comment by thadguidry** [comment]
TL:DR Any version of OpenRefine that is **pre 3.5.1** is subject to having corrupted workspaces at times depending on preference values not using UTF-8, and hence losing starred or history expressions.

OK, I think I see the issue, we did have some serialization issues where most of them all were fixed around year 2019 and one of the last ones concerning logging of expressions was addressed by @wetneb in https://github.com/OpenRefine/OpenRefine/pull/2264/files and our deserialization code not recovering from a "scripting.starred-expressions":null from corrupted workspaces as noted by @tfmorris in https://github.com/OpenRefine/OpenRefine/issues/2544#issuecomment-629375199

So, I think this issue of still losing history or starred GREL expressions is likely only for those that might be upgrading from a 3.0-3.1 release to a >= 3.2 release.
And where all of the serialization issues and preferences (UTF-8) and expression history/starred eventually got resolved by 3.5-beta1 release as noted within the various comments in the commits for: https://github.com/OpenRefine/OpenRefine/pull/2657

So ultimately, I don't know what the real fix should be here for those on OpenRefine 3.0-3.1 and wanting to upgrade to >= 3.2 release, but perhaps with my testing and detailed notes above @wetneb has more details to rationalize about a fix or not. ;-)

**Comment by thadguidry** [comment]
@wetneb Going to close this issue as I think it is resolved now.
